### ISSUE TITLE
Provide extension option to enable strictTemplates (override compiler options from tsconfig)

### ISSUE THREAD

**Issue by atscott** [angular/vscode-ng-language-service#1418]
With the release of v12, the Ivy-based Language Service became the default. With that, the information provided by the extension follows what the application has configured the compiler to interpret (see Alex's [blog post](https://blog.angular.io/under-the-hood-of-the-language-service-ab763c26f522) for more information on why this was done). As a result, developers upgrading from previous versions that have apps which do not have strict templates enabled often experience a loss of information from the extension. It also may not be feasible for large projects to enable `strictTemplates` immediately due to an abundance of new errors that would need to be fixed.

This feature request is to provide an extension option to override the project's `angularCompilerOptions`. This would allow the language service to enable `strictTemplates`. 

Importantly, this would likely result in the language service producing diagnostics that are not produced when running/compiling the application. However, this may still be a more ideal state than the current experience (which would be simply not getting _any_ useful information for lots of locations in the template which require stricter type checking options).

**Comment by atscott** [comment]
Closing due to lack of interest

**Comment by Zuckjet** [comment]
:thumbsup: it is very useful, hope this idea will land as soon as possible.

**Comment by atscott** [comment]
@Zuckjet I'm a bit curious what your expected outcome is. These type failures that happen because of `strictTemplates: true` produce type errors in the project but would not prevent you from developing the application. `ng serve` will still serve the app with updates despite type checking errors. 

Why not just set `strictTemplates: "true"` in your local `tsconfig.json` but not commit the changes to the repository?

This feature would create a situation where the errors and information shown in `VSCode` do not match the compilation of the production app. The current behavior of the extension matches the behavior of other similar systems. In TypeScript, if you have `strict: false` in your `tsconfig.json`, the language service information reflects that and you do not have the ability to change this for the language service extension alone.

**Comment by Zuckjet** [comment]
> These type failures that happen because of strictTemplates: true produce type errors in the project but would not prevent you from developing the application. ng serve will still serve the app with updates despite type checking errors.

It seems these type errors will stop me from developing the application. right now if type errors exists, my project's html page will show:
![image](https://user-images.githubusercontent.com/17976139/142696943-cbc2039d-dd91-4b48-9e5f-a95587496fb8.png)

the type error will add a overlay on html page, which will stop me from developing my project. @atscott

**Comment by atscott** [comment]
@Zuckjet Without a specific example, I'm not able to come up with why this would happen off the top of my head. I introduced some type errors for inputs in my own project and they didn't prevent the page from loading, though they did result in the errors being shown by `ng serve` still. Given that your case did, there's also no guarantee that the language service compilation would complete either. So even if we were to add this feature, I can't say that it would solve your problem.

**Comment by Zuckjet** [comment]
@atscott I have created a mini [repo](https://github.com/Zuckjet/vscode-ng-language-service-test) for reproduce.

When you type `npm start` it will show compile error, and I cannot visit page because these type error.
![image](https://user-images.githubusercontent.com/17976139/142700373-920fb078-0704-4c08-a29e-0794cd9269ab.png)

I know these errors is not relevant with **vscode-ng-language-service** extension, it is angular compiler that complains.

What I expect is that: when I set `strictTemplates: false`, I still can use "go to definition" ability offered by this extension.

**Comment by Zuckjet** [comment]
I think I have understand what you mean. Do you mean "If angular compiler complains errors, the language service compilation will fail too" ?

I thought we can do this:  even though I set `strictTemplates: false` in my project, this extensions can still keep `strictTemplates: true` on background.  It seems it is my understand mistake, I am not familiar with vscode extension developing. If we cannot do this, It is up to you to decide whether to close this feature request, Thanks.

**Comment by atscott** [comment]
> I think I have understand what you mean. Do you mean "If angular compiler complains errors, the language service compilation will fail too" ?

That's correct. The language service directly uses the Angular compiler to interpret the project. So it's possible that if your project completely fails to compile with `npm start`, the result might be the same for the extension. The severity of each issue might be quite different though, as we've found here. 

We do strongly recommend setting `strictTemplates: true`. Not doing so can result in undetected errors in your templates and even whole portions of it can go _completely_ unchecked. Your project might be using `checkTemplateBodies: false` (this would be the case if you have no `angularCompilerOptions` set at all or do not have either `fullTemplateTypeCheck: true` or `strictTemplates: true` in there. This being disabled would mean everything inside a structural directive will be totally unchecked.

**Comment by atscott** [comment]
@Zuckjet are you able to first start your server, _then_ add `strictTemplates: true` After your server has already started successfully? In my own testing, this seemed to do the trick.

**Comment by Zuckjet** [comment]
> @Zuckjet are you able to first start your server, _then_ add `strictTemplates: true` After your server has already started successfully? In my own testing, this seemed to do the trick.

Yes, this did work, I will use this solution.

What's more, For angular version 13. I found that the error(error TS2554: Expected 0 arguments, but got 1.) which I showed in this [comment[(https://github.com/angular/vscode-ng-language-service/issues/1418#issuecomment-974522250) will not disappear even though `strictTemplates: false`.  But for angular version 12, the error will disappear if i set `strictTemplates: false` 

Anyway, I have found solution to fix my problem as you suggested, many thanks.
